### PR TITLE
Improve error/rejection messages

### LINKED ISSUE #67: Composition failure for an import refs to type of the import, not the contract type
Had the following:

``` C# 
        [ImportingConstructor]
        public StartupProjectRegistrar(
            [Import(typeof(IProjectGuidService))]IProjectGuidService2 projectGuidService,
        {
        }
```

Got the following error:

```
Microsoft.VisualStudio.ProjectSystem.VS.Debug.StartupProjectRegistrar.ctor(projectGuidService): expected exactly 1 export of Microsoft.VisualStudio.ProjectSystem.VS.IProjectGuidService2 but found 2.
    Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.VCProjectShim
    Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode
   part definition Microsoft.VisualStudio.ProjectSystem.VS.Debug.StartupProjectRegistrar
```

Expected:

```
Microsoft.VisualStudio.ProjectSystem.VS.Debug.StartupProjectRegistrar.ctor(projectGuidService): expected exactly 1 export of Microsoft.VisualStudio.ProjectSystem.VS.IProjectGuidService but found 2.
    Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.VCProjectShim
    Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode
   part definition Microsoft.VisualStudio.ProjectSystem.VS.Debug.StartupProjectRegistrar
```

Should be referring to the contract type of the import, not the identity type.

### LINKED ISSUE #68: Error when encountering a null ImportMany "collection" should be a little more obvious
Had the following code:

```
        public UnconfiguredProjectContextProvider()
        {
        }

        [ImportMany]
        public OrderPrecedenceImportCollection<IProjectGuidService> ProjectGuidServices
        {
            get;
        }
```

Notice I hadn't instantiated the collection above, and got the following:

```
Microsoft.VisualStudio.Composition.CompositionFailedException: Import "Microsoft.VisualStudio.ProjectSystem.VS.LanguageServices.UnconfiguredProjectContextProvider.ProjectGuidServices" could not be satisfied. ---> Microsoft.VisualStudio.Composition.CompositionFailedException: Unable to instantiate custom import collection type.
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.GetValueForImportSite(RuntimePartLifecycleTracker importingPartTracker, RuntimeImport import)
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.SatisfyImports()
   --- End of inner exception stack trace ---
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.SatisfyImports()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.SatisfyImmediateImports()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(PartLifecycleState nextState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose()
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.<>c__DisplayClass15_0.<GetExportedValueHelper>b__0()
   at Microsoft.VisualStudio.Composition.DelegateServices.<>c__DisplayClass2_0`1.<As>b__0()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()

```

It would be helpful to provide a little more info here - you can't instantiate a new version of the collection here, but that's not the problem in this case - we forgot to instantiate it.

### LINKED ISSUE #69: Putting an ImportMany on a non-collection type, results in InvalidOperationException
From: https://github.com/dotnet/roslyn/issues/25212.

``` VB

Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Snippets
    ' HACK: The Export attribute (As ISnippetInfoService) is used by EditorTestApp to create this
    ' SnippetInfoService on the UI thread.
    <Export(GetType(ISnippetInfoService))>
    <ExportLanguageService(GetType(ISnippetInfoService), LanguageNames.VisualBasic), [Shared]>
    Friend Class VisualBasicSnippetInfoService
        Inherits AbstractSnippetInfoService
 
        <ImportingConstructor>
        Public Sub New(serviceProvider As SVsServiceProvider, <ImportMany> listenerProvider As IAsynchronousOperationListenerProvider)
            MyBase.New(serviceProvider, Guids.VisualBasicDebuggerLanguageId, listenerProvider)
        End Sub
    End Class
End Namespace
```

```
Microsoft.VisualStudio.Composition.PartDiscoveryException: Failure while scanning type "Microsoft.VisualStudio.LanguageServices.VisualBasic.Snippets.VisualBasicSnippetInfoService". ---> System.InvalidOperationException: Sequence contains no elements
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Microsoft.VisualStudio.Composition.PartDiscovery.GetElementTypeFromMany(Type type)
   at Microsoft.VisualStudio.Composition.PartDiscovery.GetTypeIdentityFromImportingType(Type type, Boolean importMany)
   at Microsoft.VisualStudio.Composition.AttributedPartDiscovery.TryCreateImportDefinition(Type importingType, ICustomAttributeProvider member, ImmutableHashSet`1 importConstraints, ImportDefinition& importDefinition)
   at Microsoft.VisualStudio.Composition.AttributedPartDiscovery.CreateImport(ParameterInfo parameter, ImmutableHashSet`1 importConstraints)
   at Microsoft.VisualStudio.Composition.AttributedPartDiscovery.CreatePart(Type partType, Boolean typeExplicitlyRequested)
   at Microsoft.VisualStudio.Composition.PartDiscovery.CombinedPartDiscovery.CreatePart(Type partType, Boolean typeExplicitlyRequested)
   at Microsoft.VisualStudio.Composition.PartDiscovery.<>c__DisplayClass27_0.<CreateDiscoveryBlockChain>b__0(Type type)
   --- End of inner exception stack trace 
```

### PR DESCRIPTION
**Pull request by AArnott**
Fixes #67
Fixes #68
Fixes #69

### CODE REVIEW

File: src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf [hunk 1]
```diff
@@ -112,12 +113,16 @@
           <target state="translated">Ke každému importu musí být právě jedna položka.</target>
         </trans-unit>
         <trans-unit id="ExpectedExactlyOneExportButFound" translate="yes" xml:space="preserve">
-          <source>{0}: expected exactly 1 export of {1} but found {2}.{3}</source>
-          <target state="translated">{0}: očekával se právě 1 export {1}, ale počet nalezených exportů je: {2}.{3}</target>
+          <source>{0}: expected exactly 1 export matching constraints:
+{1}but found {2}.{3}</source>
```
Comment (jepetty) [inline src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf]: I think we lost a space here between '{0}' and 'but'

File: src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf [hunk 2]
```diff
@@ -112,12 +113,16 @@
...
+          <target state="needs-review-translation">{0}: očekával se právě 1 export {1}, ale počet nalezených exportů je: {2}.{3}</target>
+          <note from="MultilingualUpdate" annotates="source" priority="2">Please verify the translation’s accuracy as the source string was updated after it was translated.</note>
         </trans-unit>
         <trans-unit id="ExpectedOneOrZeroExportsButFound" translate="yes" xml:space="preserve">
-          <source>{0}: expected 1 or 0 exports of {1} but found {2}.{3}</source>
-          <target state="translated">{0}: očekával se 1 nebo 0 exportů {1}, ale počet nalezených exportů je: {2}.{3}</target>
+          <source>{0}: expected 1 or 0 exports matching constraints:
+{1}but found {2}.{3}</source>
```
Comment (jepetty) [inline src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf]: Same here

File: src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf (hunk 1 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf]: This is intentional. `{1}` gets replaced with a series of lines that end in a newline character, so the space that _so wants to be there_ would actually cause ` but found` to appear to have an extra space at the start of its new line.

File: src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf (hunk 2 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.Composition.LocalizationShell/MultilingualResources/Microsoft.VisualStudio.Composition.LocalizationShell.cs.xlf]: Same answer.

**Pull request merged**